authoritative: debian packaging update for 4.4

zeha opened the pull request:
### Short description
"backport" of #9964 for auth 4.4.

jessie/xenial is left untouched.

### Checklist
<!-- please indicate if any of these things are done/included with this Pull Request. Not all boxes need to be checked for the Pull Request to be accepted -->
I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [ ] compiled this code
- [ ] tested this code
- [ ] included documentation (including possible behaviour changes)
- [ ] documented the code
- [ ] added or modified regression test(s)
- [ ] added or modified unit test(s)
- [ ] checked that this code was merged to master - not yet, see #9964

Habbie commented:
stretch:

```
Step 42/43 : RUN PDNS_TEST_NO_IPV6=1 builder/helpers/build-debs.sh pdns-${BUILDER_VERSION}
 ---> Running in 99ec43da55b8
/pdns/pdns-4.4.0.11.debianupdate44.g155192c50 /pdns
dh_testdir
dh_testroot
dh_prep
dh_testdir
dh_testroot
dh_install
dh_install: Compatibility levels before 9 are deprecated (level 7 in use)
dh_installdocs
dh_installdocs: Compatibility levels before 9 are deprecated (level 7 in use)
dh_installchangelogs
dh_compress
dh_fixperms
dh_installdeb
dh_installdeb: Compatibility levels before 9 are deprecated (level 7 in use)
dh_gencontrol
dh_md5sums
dh_builddeb
dpkg-deb: building package 'pdns-build-deps' in '../pdns-build-deps_1.0_amd64.deb'.

The package has been created.
Attention, the package has been created in the current directory,
not in ".." as indicated by the message above!
Selecting previously unselected package pdns-build-deps.
(Reading database ... 26884 files and directories currently installed.)
Preparing to unpack pdns-build-deps_1.0_amd64.deb ...
Unpacking pdns-build-deps (1.0) ...
Reading package lists...
Building dependency tree...
Reading state information...
Correcting dependencies...Starting pkgProblemResolver with broken count: 1
Starting 2 pkgProblemResolver with broken count: 1
Investigating (0) pdns-build-deps:amd64 < 1.0 @iU mK Nb Ib >
Broken pdns-build-deps:amd64 Depends on debhelper-compat:amd64 < none @un H > (= 10)
  Removing pdns-build-deps:amd64 because I can't find debhelper-compat:amd64
Done
 Done
Starting pkgProblemResolver with broken count: 0
Starting 2 pkgProblemResolver with broken count: 0
Done
The following packages will be REMOVED:
  pdns-build-deps
0 upgraded, 0 newly installed, 1 to remove and 0 not upgraded.
1 not fully installed or removed.
After this operation, 9216 B disk space will be freed.
(Reading database ... 26888 files and directories currently installed.)
Removing pdns-build-deps (1.0) ...
mk-build-deps: Unable to install pdns-build-deps at /usr/bin/mk-build-deps line 402.
mk-build-deps: Unable to install all build-dep packages
Full build logs can be found in  /home/peter/projects/powerdns/pdns-trees/docker/builder/tmp/build_debian-stretch_4.4.0.11.debianupdate44.g155192c50_20210118-103037.log 
ERROR: Build failed
```

Habbie commented:
A selection of observed changes in the buster packaging (plus my opinion):
* pkcs11 is gone (I'm neutral on this)
* init.d supports is gone (good)
* `/var/lib/powerdns` is gone (undecided) ah, but pdns-backend-lmdb creates it because it needs it (nice)
* security-poll-suffix is set to empty (wrong for upstream)
* setuid/setgid are set to pdns (I need to check if this is right for how we use systemd)
* pdns-tools, new binaries: dnspcap2calidns+manpage (good), stubquery+no manpage (not sure, but fine with me)
* copyright file (only checked in one package, I trust it's identical in all) looks mostly better, except you made your own name longer in one spot
* `Priority' is now `optional` instead of `extra` (I don't know what this implies)
* gsqlite3 [and probably others]: the example .conf is moved into `/usr/share/doc`, and the schema files have moved the other way (I remember discussing this but I don't recall the outcome - in any case this is good for Docker users) [note that `schema.*.sql` (except bind) now lives in both places! - is that to keep existing instructions on the Internet useful? ]
* ldap: pdns-domaininfo schema added (good)

End of comparison. I will do actual testing soon.

zeha commented:
> security-poll-suffix is set to empty (wrong for upstream)

fixed

> setuid/setgid are unset [was pdns] (I need to check if this is right for how we use systemd)

I believe this is correct for systemd

> pdns-tools, new binaries: dnspcap2calidns+manpage (good), stubquery+no manpage (not sure, but fine with me)
copyright file (only checked in one package, I trust it's identical in all) looks mostly better, except you made your own name longer in one spot

fixed

> Priority is now optional instead of extra (I don't know what this implies)

"extra" is gone from Debian :) 

> gsqlite3 [and probably others]: the example .conf is moved into /usr/share/doc, and the schema files have moved the other way (I remember discussing this but I don't recall the outcome - in any case this is good for Docker users) [note that schema.*.sql (except bind) now lives in both places! - is that to keep existing instructions on the Internet useful? ]

The files in /usr/share/doc should actually be symlinks. Apart from the mssql file, which was duplicate, and I've removed.

Notes:
- I haven't tested the odbc change, will do later.
- Can you build with debhelper from backports for stretch?

Habbie commented:
> Can you build with debhelper from backports for stretch? /cc @pieterlexis

If that doesn't require the user (of the packages) to enable backports, works for me!

pieterlexis commented on builder-support/debian/authoritative/debian-buster/README.source:
```diff
@@ -1 +1,58 @@
-See /usr/share/doc/quilt/README.source
```
technically, none of this is true for the PowerDNS builds of the package. Although I'm unsure if it warrants the deviation from downstream.

zeha replied:
In Debian I've added a completely new README.source; none of that in there is correct for the upstream builds. I'd keep it out for now.

zeha replied:
The new file is really all wrong; I've removed it.

pieterlexis commented on builder-support/debian/authoritative/debian-buster/config/gmysql.conf:
```diff
@@ -0,0 +1,68 @@
+# gmysql-dbname	Database name to connect to
+#
+# gmysql-dbname=powerdns
+gmysql-dbname=_DBNAME_
```
this smells of db-config

zeha replied:
it was, but right now its used as the sed template for the autopkgtest. Somehow having separate files (one example file, one autopkgtest file) might be cleaner, but its what we have today.

pieterlexis commented on builder-support/debian/authoritative/debian-buster/config/ixfrdist.yml:
```diff
@@ -0,0 +1,102 @@
+# ixfrdist example configuration file.
```
Would it make sense to `sed` the example config into this at build-time?

zeha replied:
i had a look when introducing the file in debian, and it was too awkward, cannot remember the details.

What I'd ideally want is to tell ixfrdist "please use your good defaults, add these settings, and write the config file text to stdout".

pieterlexis replied:
> What I'd ideally want is to tell ixfrdist "please use your good defaults, add these settings, and write the config file text to stdout".

When we introduce a new config parser, we'll do that

pieterlexis commented on builder-support/debian/authoritative/debian-buster/control:
```diff
@@ -1,40 +1,30 @@
-               libluajit-5.1-dev [!arm64],
-               liblua5.3-dev [arm64],
+               libluajit-5.1-dev [amd64] | liblua5.1-0-dev,
+               liblua5.1-0-dev [!amd64],
```
does this move to plain Lua on non-AMD64?

Habbie replied:
yes - kpfleming did that for the ARM64 case

Habbie replied:
oh, but there's a downgrade from 5.3 to 5.1 here for non-amd64!

zeha replied:
So, this should do:
- on amd64, force luajit
- on other archs, use luajit if its there
- then use lua 5.1. 

I figured its smarter to use 5.1 across all packages and products. Can certainly put 5.3 back in.

pieterlexis replied:
> I figured its smarter to use 5.1 across all packages and products. Can certainly put 5.3 back in.

@Habbie, your call

Habbie replied:
5.3 please

pieterlexis commented on builder-support/debian/authoritative/debian-buster/control:
```diff
@@ -1,40 +1,30 @@
                libsqlite3-dev,
                libssl-dev,
                libsystemd-dev [linux-any],
-               libtool,
```
this is pulled in via another dep?

zeha replied:
via dh-autoreconf, yes.

pieterlexis commented on builder-support/debian/authoritative/debian-buster/copyright:
```diff
@@ -52,20 +58,32 @@ Files: modules/ldapbackend/* pdns/zone2ldap.cc
-Files: ext/yahttp/*
-Copyright: 2014 Aki Tuomi
-License: Expat
+Files: ext/incbin/*
```
incbin is gone on master btw.

zeha replied:
i've noticed. for 4.5 :-)

pieterlexis commented on builder-support/debian/authoritative/debian-buster/pdns-backend-ldap.install:
```diff
@@ -1,3 +1,4 @@
-modules/ldapbackend/dnsdomain2.schema etc/ldap/schema/
 usr/bin/zone2ldap usr/bin/
 usr/lib/*/pdns/libldapbackend.so*
+usr/share/doc/pdns/dnsdomain2.schema etc/ldap/schema/
```
would it make sense to name this pdns-dnsdomain2.schema?

zeha replied:
maybe, but then it should be named like that in the tarball already

pieterlexis replied:
@Habbie opinions?

Habbie replied:
it's not a pdns-specific file; let's keep it as is

pieterlexis commented on builder-support/debian/authoritative/debian-buster/pdns-backend-lmdb.dirs:
```diff
@@ -0,0 +1 @@
+var/lib/powerdns
```
given the fact that the gsqlite3 and bind backends _want_ to use this dir (due to systemd's ProtectSystem setting), would it make sense to declare this dir in the pdns-server package?

zeha replied:
pdns-backend-bind and pdns-backend-gsqlite3 both have the same dirs file, so they will create it. I don't think we should move that into pdns-server.

pieterlexis commented on builder-support/debian/authoritative/debian-buster/pdns-backend-mysql.NEWS:
```diff
@@ -0,0 +1,9 @@
+pdns (4.1.3-4) unstable; urgency=medium
```
I'm not sure we should ship this in our packages, idem for the ones in gpgsql and gsqlite3

zeha replied:
your call :)

pieterlexis replied:
let's drop them then :)

zeha replied:
should I also drop pdns-server.NEWS?

pieterlexis replied:
yes please

zeha replied:
done!

pieterlexis commented on builder-support/debian/authoritative/debian-buster/pdns-server.lintian-overrides:
```diff
@@ -1,4 +1,2 @@
-pdns-server: non-standard-file-perm etc/powerdns/pdns.conf 0600 != 0644
-# We load lsb-functions conditionally.
-pdns-server: init.d-script-needs-depends-on-lsb-base
+pdns-server: non-standard-file-perm etc/powerdns/pdns.conf 0640 != 0644
```
0640 is a fine permission to have though?

zeha replied:
yes :) it just complains because "non-standard"

pieterlexis commented on builder-support/debian/authoritative/debian-buster/rules:
```diff
@@ -1,92 +1,68 @@
-# Backends
-backends := bind ldap pipe gmysql godbc gpgsql gsqlite3 geoip lmdb lua2 remote random tinydns
+# Vendor and version
+CXXFLAGS += -DPACKAGEVERSION='"$(DEB_VERSION).$(DEB_VENDOR)"'
```
I believe this is not what we want in our packages (for secpoll)

pieterlexis commented on builder-support/debian/authoritative/debian-buster/rules:
```diff
@@ -1,92 +1,68 @@
-		--enable-experimental-pkcs11 \
 		--enable-lua-records \
-		--enable-ixfrdist \
+		--enable-reproducible \
```
we had this disabled in our packages before, so we knew the exact build info.

zeha replied:
fixed

pieterlexis commented on builder-support/debian/authoritative/debian-buster/rules:
```diff
@@ -1,92 +1,68 @@
 override_dh_auto_test:
 ifeq (,$(filter nocheck,$(DEB_BUILD_OPTIONS)))
+	./debian/tests-source/run
 	echo Skipping make test, as dependencies are missing.
```
it says it is skipping, but then proceeds to run `make check`?

zeha replied:
make check, but not make test. IIRC there was an important difference, at least previously.

pieterlexis replied:
```
make: *** No rule to make target 'test'.  Stop.
```

:man_shrugging:

pieterlexis commented on builder-support/debian/authoritative/debian-buster/tests/supported-algos:
```diff
@@ -10,7 +10,7 @@ trap cleanup EXIT
 pdnsutil list-algorithms 2>/dev/null | tee "$TMPFILE"
 
-for algo in RSASHA1 RSASHA1-NSEC3-SHA1 RSASHA256 RSASHA512 ECC-GOST ECDSAP256SHA256 ECDSAP384SHA384 ED25519; do
+for algo in RSASHA1 RSASHA1-NSEC3-SHA1 RSASHA256 RSASHA512 ECDSAP256SHA256 ECDSAP384SHA384 ED25519; do
```
I _think_ `ED448` (algo 16) is supported via OpenSSL as well.

Habbie replied:
15+16 is a package deal in openssl

Habbie replied:
we either have both or neither

zeha replied:
added ED448

pieterlexis reviewed (commented):
A bunch of comments and some questions phrased as comments (sorry). No serious things that would prevent merging as-is. I would like to run this through our buildbot at least once to test this builds on Ubuntu as well.

pieterlexis commented:
Builds of commit 1613283:

* https://builder.powerdns.com/#/builders/99/builds/1657
* https://builder.powerdns.com/#/builders/113/builds/861

Habbie commented:
> The files in /usr/share/doc should actually be symlinks.

Ah! `pkgdiff` said `/usr/share/doc/pdns-backend-sqlite3/schema.sqlite3.sql` was unchanged, but that's wrong. The eventual content might be the same, but a symlink is not 'unchanged'.

Habbie commented:
Why is lmdb's conf in /etc instead of somewhere in /usr like all the others? (except bind but that's not a new change)

Habbie commented:
And, I said I was neutral on PKCS11 going away, but what's the reason it's gone?

Habbie commented:
Other than those two questions, this looks excellent, thanks!

jsoref commented on builder-support/debian/authoritative/debian-buster/config/ldap.conf:
```diff
@@ -1,7 +1,77 @@
+# ldap-filter-lookup=(:target:)
+
+#################################
+# ldap-host	One or more LDAP server with ports or LDAP URIs (separated by spaces)
```
```suggestion
# ldap-host	One or more LDAP servers with ports or LDAP URIs (separated by spaces)
```

zeha replied:
this is auto-generated from pdns code, please fix it there first, or it will be reintroduced.

jsoref reviewed (commented):
I know this isn't quite the right place for these comments, but this is a rare chance to see a bunch of related files at once, so if people could think about/discuss these for changing in their upstream files :-)

Habbie commented:
@jsoref thanks - going to perhaps ignore it here but we can fix it on master.

pieterlexis commented:
@zeha can you answer these 2 questions?

> Why is lmdb's conf in /etc instead of somewhere in /usr like all the others? (except bind but that's not a new change)

> And, I said I was neutral on PKCS11 going away, but what's the reason it's gone?

zeha commented:
> @zeha can you answer these 2 questions?
> 
> > Why is lmdb's conf in /etc instead of somewhere in /usr like all the others? (except bind but that's not a new change)

lmdb just works (like bind). If you install the pdns-backend-lmdb package, you immediately have a running config. For all the other backends, you need to install and setup stuff. But we could -not- do that.

> > And, I said I was neutral on PKCS11 going away, but what's the reason it's gone?

Apparently b1de17d77e1dff71b77d57efb9961f1fd7f5d8a3 never made it into Debian's packaging. No other reason.

Habbie commented:
> lmdb just works (like bind). If you install the pdns-backend-lmdb package, you immediately have a running config. For all the other backends, you need to install and setup stuff. But we could -not- do that.

This makes sense to me. @pieterlexis?

> Apparently [b1de17d](https://github.com/PowerDNS/pdns/commit/b1de17d77e1dff71b77d57efb9961f1fd7f5d8a3) never made it into Debian's packaging. No other reason.

Ah, a two way sync problem :) Do you want to take that change to your side?

zeha commented on builder-support/debian/authoritative/debian-buster/config/ixfrdist.yml:
```diff
@@ -0,0 +1,102 @@
+#failed-soa-retry: 30
+
+# Whether record compression should be enabled, leading to smaller answers
+# at the cost of an increased CPU and memory usage.
```
these texts came from docs/manpages/ixfrdist.yml.5.rst, please fix it there first.

Habbie replied:
@jsoref is aware, I allowed him to just leave it here for now

zeha commented on builder-support/debian/authoritative/debian-buster/config/ixfrdist.yml:
```diff
@@ -0,0 +1,102 @@
+# Whether record compression should be enabled, leading to smaller answers
+# at the cost of an increased CPU and memory usage.
+#
+#compress: false
```
they are however, completely different file formats, and are written by different pieces of software. this is not introduced in this PR, so please fix this elsewhere first, otherwise it will always be reintroduced.

zeha commented:
> > Apparently [b1de17d](https://github.com/PowerDNS/pdns/commit/b1de17d77e1dff71b77d57efb9961f1fd7f5d8a3) never made it into Debian's packaging. No other reason.
> 
> Ah, a two way sync problem :) Do you want to take that change to your side?

Pushed (here, and for Debian).

zeha commented:
Is this the last thing holding up 4.4.1?

For Debian, I'd need to put an upload in before Feb 8 or 9. (I know I'm part of the delay, sorry.)

pieterlexis commented:
> This makes sense to me. @pieterlexis?

aye

Habbie commented:
> Is this the last thing holding up 4.4.1?

There's one other PR but that just needs a quick review from me.
